lib/rules/no-constant-condition.js

```diff
@@ -207,6 +207,8 @@ module.exports = {
             "ForStatement:exit": checkConstantConditionLoopInSet,
             FunctionDeclaration: enterFunction,
             "FunctionDeclaration:exit": exitFunction,
+            FunctionExpression: enterFunction,
+            "FunctionExpression:exit": exitFunction,
```

Comment:
Should this also include `ArrowFunctionExpression`? It seems suspicious to me that we have logic trying to keep track of the current function and we don't check change the state when we encounter an arrow function. However, I'm not sure if this results in any observable bugs. (Generator functions can't be arrow functions, but it seems like there might be a problem if there is an arrow function within a generator function, because the "current function" would be incorrect.)

Reply:
@not-an-aardvark good question. I traced the logic a bit. The only thing `loopsInCurrentScope` really "does" is to be cleared when a `YieldExpression` is encountered, other than that it's just a holding ground for reportable loops (loops get added to `loopsInCurrentScope` when they get visited on entry, and then reported if they're still there on exit)

So while agreed it's a suspicious definition of "scope", as long as the invariant holds that we can't encounter a `YieldExpression` from a different function while still in the same "scope" then I think the logic is OK. I believe that now that `FunctionExpression` triggers a new "scope", that invariant is valid, since you don't appear to be able to `yield` from inside an `ArrowFunctionExpression`

Reply:
Ok, sounds reasonable to me.